### PR TITLE
ctc_loss odds and ends

### PR DESCRIPTION
**Pull request by t-vi**
- Add convenience wrapper to pass tensors as input_lengths, target_lengths
- Fix documentation example
- Check BLANK >= 0

Thank you, Simon and Soumith for the suggestions!

### CODE REVIEW

**Review by soumith — APPROVED** [review]

**Review by soumith — CHANGES_REQUESTED** [review]
can you add the allowed types for each input.

For example:

```
input_lengths: Tensor or tuple of size :math:`(N)`.
```

File: aten/src/ATen/native/native_functions.yaml [hunk 1]
```diff
@@ -512,6 +512,9 @@
 - func: ctc_loss(Tensor log_probs, Tensor targets, IntList input_lengths, IntList target_lengths, int64_t blank=0, int64_t reduction=Reduction::ElementwiseMean) -> Tensor
   variants: function
 
+- func: ctc_loss(Tensor log_probs, Tensor targets, Tensor input_lengths, Tensor target_lengths, int64_t blank=0, int64_t reduction=Reduction::ElementwiseMean) -> Tensor
```
Comment (SsnL) [inline aten/src/ATen/native/native_functions.yaml]: worth a comment here :)

File: torch/nn/modules/loss.py [hunk 2]
```diff
@@ -1135,15 +1135,15 @@ class CTCLoss(_Loss):
...
-        log_probs: :math:`(T, N, C)` where `C = number of characters in alphabet including blank`,
+        log_probs: Tensor of size :math:`(T, N, C)` where `C = number of characters in alphabet including blank`,
             `T = input length`, and `N = batch size`.
             The logarithmized probabilities of the outputs
             (e.g. obtained with :func:`torch.nn.functional.log_softmax`).
-        targets: :math:`(N, S)` or `(sum(target_lenghts))`.
+        targets: Tensor of size :math:`(N, S)` or `(sum(target_lenghts))`.
             Targets (cannot be blank). In the second form, the targets are assumed to be concatenated.
-        input_lengths: :math:`(N)`.
+        input_lengths: Tuple or tensor of size :math:`(N)`.
```
Comment (SsnL) [inline torch/nn/modules/loss.py]: a minor inconsistency here: actually `list` also works.

File: torch/nn/modules/loss.py [hunk 3]
```diff
@@ -1152,8 +1152,8 @@ class CTCLoss(_Loss):
         >>> ctc_loss = nn.CTCLoss()
         >>> log_probs = torch.randn(50, 16, 20).log_softmax(2).detach().requires_grad_()
         >>> targets = torch.randint(1, 21, (16, 30), dtype=torch.long)
-        >>> input_lengths = torch.full((16,), 50, dtype=torch.long)
-        >>> target_lengths = torch.randint(10,30,(16,), dtype=torch.long)
+        >>> input_lengths = torch.full((16,), 50, dtype=torch.long).toarray()
```
Comment (SsnL) [inline torch/nn/modules/loss.py]: hmm are you thinking about `tolist`? `toarray` doesn't seem to exist for me.

File: torch/nn/modules/loss.py (hunk 2 above)
Comment (SsnL) [reply torch/nn/modules/loss.py]: it's up to you to fix this or not :)

Comment (t-vi) [reply torch/nn/modules/loss.py]: I think we call it tuple often when it's an IntList on the ATen side (e.g. 2d convolution stride etc.). I'm not attached to not mentioning lists, but personally, I don't think it adds much, either.

File: torch/nn/modules/loss.py (hunk 3 above)
Comment (t-vi) [reply torch/nn/modules/loss.py]: Ouch.

File: torch/nn/modules/loss.py (hunk 2 above)
Comment (SsnL) [reply torch/nn/modules/loss.py]: I agree. It's just kinda confusing because the example below uses lists.

**Review by soumith — APPROVED** [review]
